### PR TITLE
Update dtype references when creation of new array fails

### PR DESCRIPTION
**Pull request by gdevanla**
My guess is that dtype needs to be decre'fed in this case. Therefore, I made this change. My understanding on the steal/borrow concept on this might be incorrect. 

Edit: Also, I would appreciate suggestions on how to test this change (if the change is valid)

### CODE REVIEW

File: src/_arraykit.c [hunk 1]
```diff
@@ -198,11 +198,12 @@ AK_ArrayDeepCopy(PyArrayObject *array, PyObject *memo)
                 array,
                 dtype,
                 NPY_ARRAY_ENSURECOPY);
-        if (memo) {
-            if (!array_new || PyDict_SetItem(memo, id, array_new)) {
-                Py_XDECREF(array_new);
-                goto error;
-            }
+        if (!array_new) { // we unnecessarily incref'ed the dtype
+            Py_DECREF(dtype);
```
[the remark this answers is not in the document]
Comment (gdevanla) [reply src/_arraykit.c]: I see. Is there a possibility they fail before the 'stealing' happens?

File: src/_arraykit.c [hunk 2]
```diff
@@ -198,11 +198,12 @@ AK_ArrayDeepCopy(PyArrayObject *array, PyObject *memo)
                 array,
                 dtype,
                 NPY_ARRAY_ENSURECOPY);
-        if (memo) {
-            if (!array_new || PyDict_SetItem(memo, id, array_new)) {
-                Py_XDECREF(array_new);
-                goto error;
-            }
+        if (!array_new) { // we unnecessarily incref'ed the dtype
```
Comment (brandtbucher) [inline src/_arraykit.c:201]: This comment no longer applies:

```suggestion
        if (!array_new) {
```

**Review by brandtbucher — CHANGES_REQUESTED** [review]

File: src/_arraykit.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (brandtbucher) [reply src/_arraykit.c]: No, it's their job to steal even in the failure path.

**Review by brandtbucher — APPROVED** [review]

File: src/_arraykit.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (gdevanla) [reply src/_arraykit.c]: Good point. Got it.

Is there a good reference/pattern to understand when a function steals or borrows? Seems like lot of guess work or scanning the implementation of the function.

[the remark this answers is not in the document]
Comment (brandtbucher) [reply src/_arraykit.c]: The it should always borrow, unless the docs explicitly say it steals. It's a bug (in either the docs or the implementation) if a function steals without documenting it.

[the remark this answers is not in the document]
Comment (brandtbucher) [reply src/_arraykit.c]: Same for return values: a function should always return a new reference (owned by the caller) unless it is explicitly documented as borrowed.

**Pull request merged**